File: examples/karpathy/bench.py
"""Benchmark Karpathy's micrograd across dataset sizes and architectures."""

import argparse
import gc
import json
import platform
import time
import tracemalloc
from pathlib import Path

from nn import MLP
from utils import (
    cross_entropy_loss,
    full_arch,
    hinge_loss,
    load_dataset,
    load_weights,
    update_params,
    weights_path,
)

DATASET_SIZES = [100, 200, 500, 1000, 5000, 10000]
ARCHITECTURES = [[8, 1], [16, 16, 1], [32, 32, 1], [16, 32, 16, 1]]
N_TRIALS = 5
N_STEPS = 100
STEP_COUNTS = [1, 10, 100, 500, 1000]
STEP_SWEEP_N_SAMPLES = 200
STEP_SWEEP_ARCH = [16, 16, 1]
LR = 1.0
PHASES = ("forward", "backward", "update")
ACTIVATIONS = {"relu", "tanh"}
LOSSES = {"hinge", "cross_entropy"}


def parse_int_list(value: str) -> list[int]:
    return [int(part.strip()) for part in value.split(",") if part.strip()]


def parse_activation(value: str) -> str:
    activation = value.strip().lower()
    if activation not in ACTIVATIONS:
        raise argparse.ArgumentTypeError(f"unsupported activation: {value}")
    return activation


def parse_loss(value: str) -> str:
    loss = value.strip().lower().replace("-", "_")
    if loss in {"crossentropy", "bce"}:
        loss = "cross_entropy"
    if loss not in LOSSES:
        raise argparse.ArgumentTypeError(f"unsupported loss: {value}")
    return loss


def resolve_loss_fn(loss: str):
    return hinge_loss if loss == "hinge" else cross_entropy_loss


def write_output(path: Path, output: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp_path = path.with_suffix(path.suffix + ".tmp")
    with open(tmp_path, "w") as f:
        json.dump(output, f, indent=2)
    tmp_path.replace(path)


def run_trial(
    n_samples: int,
    arch: list[int],
    n_steps: int = N_STEPS,
    lr: float = LR,
    activation: str = "relu",
    loss: str = "hinge",
) -> dict:
    X, y = load_dataset(n_samples)
    model = MLP(2, arch, activation=activation)
    load_weights(model, weights_path(arch))
    loss_fn = resolve_loss_fn(loss)

    gc.collect()

    phase_seconds = dict.fromkeys(PHASES, 0.0)
    tracemalloc.start()
    t_start = time.perf_counter()
    for k in range(n_steps):
        phase_start = time.perf_counter()
        total_loss, acc = loss_fn(model, X, y)
        phase_seconds["forward"] += time.perf_counter() - phase_start

        phase_start = time.perf_counter()
        model.zero_grad()
        total_loss.backward()
        phase_seconds["backward"] += time.perf_counter() - phase_start

        phase_start = time.perf_counter()
        update_params(model, k, n_steps, lr)
        phase_seconds["update"] += time.perf_counter() - phase_start

    elapsed = time.perf_counter() - t_start
    current_bytes, peak_bytes = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    return {
        "elapsed_seconds": elapsed,
        "seconds_per_step": elapsed / n_steps,
        "phase_seconds": phase_seconds,
        "phase_seconds_per_step": {
            phase: seconds / n_steps for phase, seconds in phase_seconds.items()
        },
        "memory": {
            "tracking": "tracemalloc",
            "current_bytes": current_bytes,
            "peak_bytes": peak_bytes,
            "total_bytes": peak_bytes,
            "bytes_per_step": peak_bytes / n_steps,
            "phase_bytes": None,
            "phase_bytes_per_step": None,
        },
        "final_loss": total_loss.data,
        "final_acc": acc,
    }


def run_config(
    n_samples: int,
    arch: list[int],
    n_steps: int = N_STEPS,
    n_trials: int = N_TRIALS,
    lr: float = LR,
    activation: str = "relu",
    loss: str = "hinge",
    section: str = "config",
    config_index: int = 1,
    total_configs: int = 1,
    quiet: bool = False,
) -> dict:
    label = (
        f"n={n_samples} arch={full_arch(arch)} steps={n_steps} "
        f"activation={activation} loss={loss}"
    )
    if not quiet:
        print("", flush=True)
        print(f"[{section} {config_index:02d}/{total_configs:02d}] {label}", flush=True)

    trials = []
    for t in range(n_trials):
        trial = run_trial(
            n_samples,
            arch,
            n_steps=n_steps,
            lr=lr,
            activation=activation,
            loss=loss,
        )
        trials.append(trial)
        if not quiet:
            print(
                f"    trial {t + 1}/{n_trials}  {trial['elapsed_seconds']:.2f}s",
                flush=True,
            )

    return {
        "n_samples": n_samples,
        "arch": full_arch(arch),
        "n_steps": n_steps,
        "lr": lr,
        "activation": activation,
        "loss": loss,
        "trials": trials,
    }


def throughput_sweep(
    results: list[dict],
    n_steps: int = N_STEPS,
    n_trials: int = N_TRIALS,
    lr: float = LR,
    activation: str = "relu",
    loss: str = "hinge",
    checkpoint=None,
    quiet: bool = False,
) -> None:
    total_configs = len(ARCHITECTURES) * len(DATASET_SIZES)
    config_index = 0

    for arch in ARCHITECTURES:
        for n_samples in DATASET_SIZES:
            config_index += 1
            results.append(run_config(
                n_samples,
                arch,
                n_steps=n_steps,
                n_trials=n_trials,
                lr=lr,
                activation=activation,
                loss=loss,
                section="throughput",
                config_index=config_index,
                total_configs=total_configs,
                quiet=quiet,
            ))
            if checkpoint is not None:
                checkpoint()


def step_sweep(
    results: list[dict],
    n_samples: int = STEP_SWEEP_N_SAMPLES,
    arch: list[int] | None = None,
    step_counts: list[int] | None = None,
    n_trials: int = N_TRIALS,
    lr: float = LR,
    activation: str = "relu",
    loss: str = "hinge",
    checkpoint=None,
    quiet: bool = False,
) -> None:
    arch = STEP_SWEEP_ARCH if arch is None else arch
    step_counts = STEP_COUNTS if step_counts is None else step_counts
    total_configs = len(step_counts)
    for config_index, n_steps in enumerate(step_counts, start=1):
        results.append(run_config(
            n_samples,
            arch,
            n_steps=n_steps,
            n_trials=n_trials,
            lr=lr,
            activation=activation,
            loss=loss,
            section="step",
            config_index=config_index,
            total_configs=total_configs,
            quiet=quiet,
        ))
        if checkpoint is not None:
            checkpoint()


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--mode",
        choices=["all", "throughput", "step-sweep", "single"],
        default="all",
        help="which benchmark section to run",
    )
    parser.add_argument("--n-samples", type=int, default=STEP_SWEEP_N_SAMPLES)
    parser.add_argument("--arch", type=parse_int_list, default=STEP_SWEEP_ARCH)
    parser.add_argument("--n-steps", type=int, default=N_STEPS)
    parser.add_argument("--step-counts", type=parse_int_list, default=STEP_COUNTS)
    parser.add_argument("--trials", type=int, default=N_TRIALS)
    parser.add_argument("--lr", type=float, default=LR)
    parser.add_argument("--activation", type=parse_activation, default="relu")
    parser.add_argument("--loss", type=parse_loss, default="hinge")
    parser.add_argument("--quiet", action="store_true")
    parser.add_argument(
        "--out",
        type=Path,
        default=Path(__file__).resolve().parent.parent / "results" / "bench_python.json",
    )
    return parser


def bench(argv: list[str] | None = None) -> None:
    args = build_parser().parse_args(argv)
    results = []
    step_sweep_results = []
    output = {
        "language": "python",
        "version": platform.python_version(),
        "platform": platform.machine(),
        "mode": args.mode,
        "n_trials": args.trials,
        "results": results,
        "step_sweep_results": step_sweep_results,
        "status": "running",
    }

    def checkpoint() -> None:
        write_output(args.out, output)

    checkpoint()

    if args.mode == "all":
        throughput_sweep(
            results,
            n_trials=args.trials,
            lr=args.lr,
            activation=args.activation,
            loss=args.loss,
            checkpoint=checkpoint,
            quiet=args.quiet,
        )
        step_sweep(
            step_sweep_results,
            n_trials=args.trials,
            lr=args.lr,
            activation=args.activation,
            loss=args.loss,
            checkpoint=checkpoint,
            quiet=args.quiet,
        )
    elif args.mode == "throughput":
        throughput_sweep(
            results,
            n_steps=args.n_steps,
            n_trials=args.trials,
            lr=args.lr,
            activation=args.activation,
            loss=args.loss,
            checkpoint=checkpoint,
            quiet=args.quiet,
        )
    elif args.mode == "step-sweep":
        step_sweep(
            step_sweep_results,
            n_samples=args.n_samples,
            arch=args.arch,
            step_counts=args.step_counts,
            n_trials=args.trials,
            lr=args.lr,
            activation=args.activation,
            loss=args.loss,
            checkpoint=checkpoint,
            quiet=args.quiet,
        )
    elif args.mode == "single":
        results.append(
            run_config(
                args.n_samples,
                args.arch,
                n_steps=args.n_steps,
                n_trials=args.trials,
                lr=args.lr,
                activation=args.activation,
                loss=args.loss,
                section="single",
                config_index=1,
                total_configs=1,
                quiet=args.quiet,
            )
        )
        checkpoint()

    output["status"] = "complete"
    checkpoint()
    if not args.quiet:
        print(f"\nSaved {args.out}")


if __name__ == "__main__":
    bench()

File: examples/jeggs/bench.jl
using Printf
using Micrograd

include(joinpath(@__DIR__, "..", "script_utils.jl"))
include("utils.jl")
using .MicrogradExampleScriptUtils
using .MicrogradExampleUtils

const DATASET_SIZES = [100, 200, 500, 1000, 5000, 10000]
const ARCHITECTURES = [[8, 1], [16, 16, 1], [32, 32, 1], [16, 32, 16, 1]]
const N_TRIALS = 5
const N_STEPS = 100
const STEP_COUNTS = [1, 10, 100, 500, 1000]
const STEP_SWEEP_N_SAMPLES = 200
const STEP_SWEEP_ARCH = [16, 16, 1]
const LR = 1.0
const PHASES = ("forward", "backward", "update")

option_activation(options, key, default) = haskey(options, key) ? activation_symbol(options[key]) : activation_symbol(default)
option_loss(options, key, default) = haskey(options, key) ? loss_symbol(options[key]) : loss_symbol(default)

function phase_dict(value)
    return Dict(phase => value for phase in PHASES)
end

# ─── Single trial ────────────────────────────────────────────────────

function run_trial(
    n_samples::Int,
    arch::Vector{Int};
    n_steps::Int=N_STEPS,
    lr::Real=LR,
    activation::Symbol=:relu,
    loss::Symbol=:hinge,
)
    X, y = load_dataset(n_samples)
    model = MLP(2, arch; activation=activation)
    load_weights!(model, weights_path(arch))
    loss_fn = loss_function(loss)

    GC.gc()

    local total_loss, acc
    phase_seconds = phase_dict(0.0)
    phase_bytes = phase_dict(0)

    start_ns = time_ns()
    for step in 0:(n_steps - 1)
        forward = @timed loss_fn(model, X, y)
        total_loss, acc = forward.value
        phase_seconds["forward"] += forward.time
        phase_bytes["forward"] += forward.bytes

        backward = @timed begin
            zero_grad!(model)
            backward!(total_loss)
        end
        phase_seconds["backward"] += backward.time
        phase_bytes["backward"] += backward.bytes

        update = @timed begin
            update_params!(model, step, n_steps, lr)
        end
        phase_seconds["update"] += update.time
        phase_bytes["update"] += update.bytes
    end
    elapsed = (time_ns() - start_ns) / 1e9

    total_bytes = sum(values(phase_bytes))

    return Dict(
        "elapsed_seconds" => elapsed,
        "seconds_per_step" => elapsed / n_steps,
        "phase_seconds" => phase_seconds,
        "phase_seconds_per_step" => Dict(
            phase => phase_seconds[phase] / n_steps for phase in PHASES
        ),
        "memory" => Dict(
            "tracking" => "timed_bytes",
            "current_bytes" => nothing,
            "peak_bytes" => nothing,
            "total_bytes" => total_bytes,
            "bytes_per_step" => total_bytes / n_steps,
            "phase_bytes" => phase_bytes,
            "phase_bytes_per_step" => Dict(
                phase => phase_bytes[phase] / n_steps for phase in PHASES
            ),
        ),
        "final_loss" => total_loss.data,
        "final_acc" => acc,
    )
end

function run_config(
    n_samples::Int,
    arch::Vector{Int};
    n_steps::Int=N_STEPS,
    n_trials::Int=N_TRIALS,
    lr::Real=LR,
    activation::Symbol=:relu,
    loss::Symbol=:hinge,
    section::AbstractString="config",
    config_index::Int=1,
    total_configs::Int=1,
    quiet::Bool=false,
)
    arch_label = full_arch(arch)
    label = "n=$n_samples arch=$arch_label steps=$n_steps activation=$activation loss=$loss"
    if !quiet
        println()
        @printf("[%s %02d/%02d] %s\n", section, config_index, total_configs, label)
    end

    trials = []
    for t in 1:n_trials
        trial = run_trial(n_samples, arch; n_steps=n_steps, lr=lr, activation=activation, loss=loss)
        quiet || @printf("    trial %d/%d  %.2fs\n", t, n_trials, trial["elapsed_seconds"])
        push!(trials, trial)
    end

    return Dict(
        "n_samples" => n_samples,
        "arch" => arch_label,
        "n_steps" => n_steps,
        "lr" => lr,
        "activation" => String(activation),
        "loss" => String(loss),
        "trials" => trials,
    )
end

# ─── Benchmark sweep ─────────────────────────────────────────────────

function throughput_sweep!(
    results::Vector{Any};
    n_steps::Int=N_STEPS,
    n_trials::Int=N_TRIALS,
    lr::Real=LR,
    activation::Symbol=:relu,
    loss::Symbol=:hinge,
    checkpoint=nothing,
    quiet::Bool=false,
)
    total_configs = length(ARCHITECTURES) * length(DATASET_SIZES)
    config_index = 0
    for arch in ARCHITECTURES
        for n_samples in DATASET_SIZES
            config_index += 1
            push!(results, run_config(
                n_samples,
                arch;
                n_steps=n_steps,
                n_trials=n_trials,
                lr=lr,
                activation=activation,
                loss=loss,
                section="throughput",
                config_index=config_index,
                total_configs=total_configs,
                quiet=quiet,
            ))
            checkpoint === nothing || checkpoint()
        end
    end
end

function step_sweep!(
    results::Vector{Any};
    n_samples::Int=STEP_SWEEP_N_SAMPLES,
    arch::Vector{Int}=STEP_SWEEP_ARCH,
    step_counts::Vector{Int}=STEP_COUNTS,
    n_trials::Int=N_TRIALS,
    lr::Real=LR,
    activation::Symbol=:relu,
    loss::Symbol=:hinge,
    checkpoint=nothing,
    quiet::Bool=false,
)
    total_configs = length(step_counts)
    for (config_index, n_steps) in enumerate(step_counts)
        push!(results, run_config(
            n_samples,
            arch;
            n_steps=n_steps,
            n_trials=n_trials,
            lr=lr,
            activation=activation,
            loss=loss,
            section="step",
            config_index=config_index,
            total_configs=total_configs,
            quiet=quiet,
        ))
        checkpoint === nothing || checkpoint()
    end
end

function bench(args=ARGS)
    options = parse_options(args)
    mode = get(options, "mode", "all")
    n_trials = option_int(options, "trials", N_TRIALS)
    n_steps = option_int(options, "n-steps", N_STEPS)
    n_samples = option_int(options, "n-samples", STEP_SWEEP_N_SAMPLES)
    arch = option_int_list(options, "arch", STEP_SWEEP_ARCH)
    lr = option_float(options, "lr", LR)
    activation = option_activation(options, "activation", :relu)
    loss = option_loss(options, "loss", :hinge)
    step_counts = option_int_list(options, "step-counts", STEP_COUNTS)
    quiet = option_bool(options, "quiet", false)
    out_path = get(options, "out", joinpath(@__DIR__, "..", "results", "bench_julia.json"))

    results = Any[]
    step_sweep_results = Any[]
    output = Dict{String,Any}(
        "language" => "julia",
        "version" => string(VERSION),
        "platform" => Sys.MACHINE,
        "mode" => mode,
        "n_trials" => n_trials,
        "results" => results,
        "step_sweep_results" => step_sweep_results,
        "status" => "running",
    )
    checkpoint() = write_json_output(output, out_path)
    checkpoint()

    if mode == "all"
        throughput_sweep!(results; n_trials=n_trials, lr=lr, activation=activation, loss=loss, checkpoint=checkpoint, quiet=quiet)
        step_sweep!(step_sweep_results; n_trials=n_trials, lr=lr, activation=activation, loss=loss, checkpoint=checkpoint, quiet=quiet)
    elseif mode == "throughput"
        throughput_sweep!(results; n_steps=n_steps, n_trials=n_trials, lr=lr, activation=activation, loss=loss, checkpoint=checkpoint, quiet=quiet)
    elseif mode == "step-sweep"
        step_sweep!(step_sweep_results;
            n_samples=n_samples,
            arch=arch,
            step_counts=step_counts,
            n_trials=n_trials,
            lr=lr,
            activation=activation,
            loss=loss,
            checkpoint=checkpoint,
            quiet=quiet,
        )
    elseif mode == "single"
        push!(results, run_config(
            n_samples,
            arch;
            n_steps=n_steps,
            n_trials=n_trials,
            lr=lr,
            activation=activation,
            loss=loss,
            section="single",
            config_index=1,
            total_configs=1,
            quiet=quiet,
        ))
        checkpoint()
    else
        error("unknown mode: $mode")
    end

    output["status"] = "complete"
    checkpoint()
    quiet || println("\nSaved $out_path")
end

if abspath(PROGRAM_FILE) == abspath(@__FILE__)
    bench()
end
